crates/flox-rust-sdk/src/models/search.rs

```diff
@@ -0,0 +1,261 @@
+    pub pname: Option<String>,
+    /// Match against the explicit version number of the package
+    pub version: Option<String>,
+    /// Match against a semver specifier
+    pub semver: Option<String>,
+    /// Match against a regular expression
+    pub r#match: Option<String>,
+}
+
+impl FromStr for Query {
+    type Err = SearchError;
+
+    // This can't actually error, but the trait requires an error type
+    fn from_str(search_term: &str) -> Result<Self, Self::Err> {
+        let q = match search_term.find('@') {
+            Some(idx) => {
+                // If we get a search term ending in '@' it most likely means the
+                // user didn't quote a search term that included a '>' character.
+                if idx >= search_term.len() - 1 {
+                    return Err(SearchError::SearchTerm(search_term.into()));
+                }
+                // Splitting at `idx` would include the `@`
+                let package = String::from(&search_term[..idx]);
+                let semver = String::from(&search_term[idx + 1..]);
+                Query {
+                    semver: Some(semver),
+                    r#match: Some(package),
+                    ..Query::default()
+                }
+            },
+            None => Query {
+                r#match: Some(search_term.into()),
+                ..Query::default()
+            },
+        };
+        Ok(q)
+    }
+}
+
+/// The deserialized search results.
+///
+/// Note that the JSON results are returned by `pkgdb` one result per line
+/// without an enclosing `[]`, so the results returned by `pkgdb` can't be
+/// directly deserialized to a JSON object. To parse the results you should
+/// use the provided `TryFrom` impl.
+#[derive(Debug, Clone, Serialize)]
+pub struct SearchResults {
+    pub results: Vec<SearchResult>,
+}
+
+impl TryFrom<&[u8]> for SearchResults {
+    type Error = SearchError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let mut results = Vec::new();
+        for maybe_line in bytes.lines() {
+            let text = maybe_line?;
+            match serde_json::from_str(&text) {
+                Ok(search_result) => results.push(search_result),
+                Err(_) => {
+                    // TODO: Errors are currently emitted to stdout as JSON, but there's no spec for the errors.
+                    //       For now if we can't turn the text into a SearchResult, we assume that it's an
+                    //       error message. If parsing that into a serde_json::Value fails, something else went
+                    //       pretty wrong.
+                    //
+                    //       Once there's a spec for the error messages we can parse this into a typed container.
+                    let err = Value::from_str(&text)?;
+                    return Err(SearchError::PkgDb(err));
+                },
+            };
+        }
+        Ok(SearchResults { results })
+    }
+}
```

Comment:
Only (nit) concern with this is that we have to read the entire output before we can parse it.

Maybe our search output is always small enough so it doesn't make a difference anyway, but if not we might consider implementing it as an inherent method that takes `impl Read` instead.

Reply:
also worth to maybe have a look at serde's own [`StreamDeserializer`](https://docs.rs/serde_json/1.0.107/serde_json/struct.StreamDeserializer.html)

Reply:
In my opinion it's clearer to just construct the list of search results this way.

Reply:
You certainly could do that, but I still think this is clearer and I don't think there's a compelling technical reason to prefer `StreamDeserializer`. Additionally, once we buffer the input it will prevent us from having to read and parse all of the input in the case that there's an error (assuming we want to stop on the first error). The exact error handling story is planned for this sprint, so I'd rather leave any further error handling discussions until then.